### ISSUE TITLE
Retro: Grouping: Spotlight

### ISSUE THREAD

**Issue by mattkrick** [ParabolInc/parabol#4952]
From SPA, we designed a pretty amazing spotlight feature:
https://www.notion.so/parabol/Big-Teams-Aviva-Matt-Aaron-7a051021f5bd42779aef38f9440d5609

We decided that this could replace breakout rooms & a few other smaller features. 
Next step is to design it. Here's the pen & ink:
<img width="535" alt="Screen_Shot_2021-03-12_at_12 20 05_PM" src="https://user-images.githubusercontent.com/5514175/117871955-df660680-b252-11eb-8ad9-ee9bca6074e3.png">

AC
- [ ] During the group phase, on each reflection card there is a "search" icon in the card footer.
- [ ] On desktop, the search icon is hidden until hovered
- [ ] Hovering the icon shows a tooltip that says "Find similar" (or whatever copy you like)
- [ ] Clicking the icon opens up the "Spotlight Modal", which is a large modal that takes up the entire meeting area (excluding left nav & right drawer)
- [ ] At the top of the spotlight modal is the card that the user clicked (it will animated into place, just like expanded groups do today)
- [ ] below the card is an autoFocused search input field with the search icon on the left
- [ ] below the input is the "SpotlightSuggestionGrid" that shows the reflection cards that are most similar to the top card. if input is empty, we still show some suggestions. 
- [ ] If no suggestions are available, then the grid says "No reflections match that search query" (copy tbd)
- [ ] If the reflection card has already been grouped, then it's group title is in the card header
- [ ] On each reflection card, the text that matches the search query is highlighted (e.g. if the search query is "cat" then the word "cat" is highlighted if it's visible on the card). Same goes for the reflection group title.
- [ ] If the reflection card is very long, then card is auto-scrolled to the first instance of the highlighted text
- [ ] The top card can be dragged & dropped onto one of the cards in the grid. Once it's dropped, the modal closes. It the drop is cancelled (it's not dropped on a specific card) then the top card animates back up to the top  
- [ ] if the top card is already is a group of 2+ cards, then it is moved out of its current group into the new one
- [ ] the bottom cards can be dragged & dropped onto the top card. if they are in a group of 2+ cards, then they are moved from their current group into the group of the top card. dropping on the top card does _not_ automatically close the modal
- [ ] Bottom cards cannot be dropped on each other (we have to draw the line somewhere. let's force interaction with that top card)

@avivapinchas @adaniels-parabol if I forgot anything, please let me know below!

@ackernaut @enriquesanchez feel free to adjust AC as you see fit

Estimate: 30 hours?

**Comment by gcrickman** [comment]
A [user](https://app.hubspot.com/contacts/3888472/contact/22796701) 🔒  mentions:

> We have a team size 10-12 people participating in the retro, when doing the retro it's hard for everyone to speak out, if you could have breakouts that would be cool

**Comment by enriquesanchez** [comment]
### TL;DR;

I recorded a little walk-through video of a prototype I made for this new feature. [You can find it here](https://www.loom.com/share/4e8f8b4e12e8417b8675e994b06d3cec).

### Find similar

> - During the group phase, on each reflection card there is a "search" icon in the card footer.
> - On desktop, the search icon is hidden until hovered
> - Hovering the icon shows a tooltip that says "Find similar" (or whatever copy you like)

When in a Group phase, users will be able to access a 'Find similar' action:

![2021-06-14 3-45-44 p m 2021-06-14 3_47_17 p m](https://user-images.githubusercontent.com/3276087/121957538-28185000-cd28-11eb-9723-099e2e9fee6b.gif)

### Spotlight modal

> - Clicking the icon opens up the "Spotlight Modal", which is a large modal that takes up the entire meeting area (excluding left nav & right drawer)
> - At the top of the spotlight modal is the card that the user clicked (it will animated into place, just like expanded groups do today)
> - below the card is an autoFocused search input field with the search icon on the left
> - below the input is the "SpotlightSuggestionGrid" that shows the reflection cards that are most similar to the top card. if input is empty, we still show some suggestions.

![imagen](https://user-images.githubusercontent.com/3276087/121958271-1daa8600-cd29-11eb-9a25-a6ffdb6df645.png)

The new Spotlight modal takes up most of the phase area, and shows the results found for the selected card. If the results include multiple cards that are already grouped together, we should only display the group once, with the highest ranking card on top.

The cards grid will take as much space needed (when there are lots of cards and plenty of space), or as little as available (when screen is small):

![imagen](https://user-images.githubusercontent.com/3276087/121958975-d2dd3e00-cd29-11eb-84a3-6db2748f6e8b.png)
_☝️ Lots of cards, big space_

![imagen](https://user-images.githubusercontent.com/3276087/121959039-e12b5a00-cd29-11eb-86e5-d64871221b29.png)
_☝️ Narrow area (mobile), so cards stack in one column_

> -  If no suggestions are available, then the grid says "No reflections match that search query" (copy tbd)

![imagen](https://user-images.githubusercontent.com/3276087/121959197-120b8f00-cd2a-11eb-8294-1df6d3098159.png)
_☝️ When no results are found we display a message and prompt the user to use the search field_

### Searching 
> - On each reflection card, the text that matches the search query is highlighted (e.g. if the search query is "cat" then the word "cat" is highlighted if it's visible on the card). Same goes for the reflection group title.

![imagen](https://user-images.githubusercontent.com/3276087/121959589-9c53f300-cd2a-11eb-82b5-bb47b571049c.png)

### Dragging and dropping
> -The top card can be dragged & dropped onto one of the cards in the grid. Once it's dropped, the modal closes. It the drop is cancelled (it's not dropped on a specific card) then the top card animates back up to the top

![2021-06-14 5-11-41 p m 2021-06-14 5_13_12 p m](https://user-images.githubusercontent.com/3276087/121966195-d1b10e80-cd33-11eb-9525-1a23e5baa97b.gif)
_☝️ The animation is a bit rudimentary given the format, but ideally we display the grouping action before the spotlight modal closes, so users are made aware of what just happened_

**Comment by avivapinchas** [comment]
Love seeing this! 

One question: When the spotlight modal is open, can I drag cards from the lower search area to the top card, or just drag the top card down? Can I drag cards in the search area around to group to each other?

**Comment by enriquesanchez** [comment]
Thanks @avivapinchas 🙌

> When the spotlight modal is open, can I drag cards from the lower search area to the top card, or just drag the top card down? Can I drag cards in the search area around to group to each other?

Yes, you can drag cards from the lower search area to the top card, and in that scenario the modal stays open to let you perform the action multiple times. That said, bottom cards cannot be dropped on each other (see AC above).

**Comment by mattkrick** [comment]
WOW. just wow. this is great! loved seeing the animations, too!

My biggest question: i see you're showing groups and individual cards in the search results. why not just individual cards? let's say there's a group of 20 cards with the group title "cats". the search query is "kitten". the 20th card says "kittens" and that's the only card that matches the search query in the group. would the user be confused as to why it showed up in the search results? is there a way to convey that the search query matches a phase on a card that is hidden in the stack? Also, clicking a group will cause the expanded group modal to open. is it weird that the big modal opens another smaller modal?

tiny nits:
- maybe change title to "Find similar reflections"
- Remove "or" from "or search for keywords..." because to emphasize that typing in the search input is the primary action
- background feel a little weird - the white cards on a white background, the header color feels so similar to the meeting background color that it feels almost unintentional, but that could just be me

**Comment by enriquesanchez** [comment]
Thanks @mattkrick 🙌

>My biggest question: i see you're showing groups and individual cards in the search results. why not just individual cards? let's say there's a group of 20 cards with the group title "cats". the search query is "kitten". the 20th card says "kittens" and that's the only card that matches the search query in the group. would the user be confused as to why it showed up in the search results? 

That's a great example, and the idea is that we show the highest ranking card at the top of the group, so in this example the group will be named "Cats" and the top-most visible card will be "kittens". 

We thought this would keep things simple and consistent, and brings awareness that the user is adding a card to an already existing group. Otherwise, I might think that I'm creating a new group?

>maybe change title to "Find similar reflections"
>Remove "or" from "or search for keywords..." to emphasize that typing in the search input is the primary action

Done, it feels much better, thank you!

![image](https://user-images.githubusercontent.com/3276087/122999999-94bccb80-d374-11eb-86e9-535648d8c8d9.png)

>background feel a little weird - the white cards on a white background, the header color feels so similar to the meeting background color that it feels almost unintentional, but that could just be me

This is a great point. I tried various color variations and kept coming back to white as the main bg color for the modal. Main reason was that behind the modal everything is already some shade of slate gray, and the visual separation between the modal and the rest of the app is important. The elevation helps, but we still need more than the drop shadow to separate such a big UI area (the modal) from the rest of the app.

And since we already use white cards on white background on other places, such as the meeting summaries, I thought it wouldn't be an issue. In this case, because cards are smaller UI elements, the elevation and/or border has a bigger visual effect when separating them from the background.

I'll spend a few more cycles on this and see if I can find an alternative. That said, I think we are in good shape to start writing implementation issues, what do you think?

**Comment by mattkrick** [comment]
yeah, it's got good guts! 
we might have some implementation issues with sorting the cards by salience. For example let's say the stack is `Cats, Cat, Kittens` and we sort it `Kittens, Cats, Cat`. Then, while you're in spotlight, I remove the `Cats` card from the stack, so it should look like a card is getting removed from the middle of your pile. If i put `Kitties` on the stack, then it should insert it in the bottom of that stack... or if the salience is high enough, it might put it on the top! If you have any strong opinions, I'd love to see em laid out in the acceptance criteria. there's gonna be a few edge cases with having 2 different sort orders that i'm probably missing.

**Comment by enriquesanchez** [comment]
Closed by implementation issue #5105
